### PR TITLE
Keep ordering of resources

### LINKED ISSUE #4950: Non-deterministic ordering of CSS resources for external resource loading.
I was just using the new external resource loading feature of bokeh. I have a series of three css files. With css the ordering is important. It seems that the order that they're actually put into the document changes on each run.

ping @canavandl - is this likely or do you need more info?

### PR DESCRIPTION
**Pull request by birdsarah**
- [x] issues: fixes #4950
- [x] tests added / passed

### CODE REVIEW

File: bokeh/resources.py [hunk 1]
```diff
@@ -290,7 +290,7 @@ def _collect_external_resources(self, resource_attr):
                 external_resources.extend(external)
 
         # Return only unique external resources
-        return list(set(external_resources))
+        return list(external_resources)
```
Comment (mattpap) [inline bokeh/resources.py]: You actually want to deduplicate resources (alphabetically or by model alphabetically). The result should be sorted in some way. You can gain some insight by looking into custom models code below.

Comment (birdsarah) [inline bokeh/resources.py]: I know we want the de-duplication, but its more important that things actually load for now.

**Comment by bryevdv** [comment]
A couple of comments:
* How hard would it be to only output resources for models that are actually in the doc being serialized, and not simply for every Model that gets imported somewhere?
* What if two models provide the same set of resources, in different and incompatible orderings?

**Comment by birdsarah** [comment]
> What if two models provide the same set of resources, in different and incompatible orderings?

Yes, I thought about that, which is why I was originally going to propose no de-duplication until we've thought it through. But at the end of the day, I don't think that's a use case that we can/should try to support.

All resources are going to end up grouped, probably at the top of a page.

Lets say you have one model with the css files: "normalize" and "style" and you need "normalize" to go before "style". 

Now lets say another model comes along and wants "style" before "normalize". In this case there's no way of supporting both models:
1) If you don't de-duplicate you now have "normalize", "style", "style", "normalize" - which will mean the css for the first model is broken
2) If you do de-duplicate you now have "normalize", "style" - which will mean the css for the second model is broken.

But lets take another example which makes the case for deduplication. You have two custom models, both use `"normalize"`*. The first model is `["normalize", "skin_1"]`, the second model is `["normalize", "skin_2"]`. In this case if we don't de-duplicate we will have `["normalize", "skin_1", "normalize", "skin_2"]` - this may break skin_1 as we will have re-run normalize after applying skin_1. But if we do de-duplicate and have `["normalize", "skin_1", "skin_2"]`.

We will need feedback from real users on how to make this work really well, but in the meantime I think the most important thing is to provide deterministic behavior so that users can easily debug this.

We will:
- load models in a deterministic order (still to do)
- keep resources in the order specified by their entries 
- de-duplicate resources, keeping the first entry of any resource
- we should document this very heavily so users can debug

----
* as an aside custom models with normalize in could be an issue for us as the may override bokeh styling - I think we're good because we're namespaced but maybe something to keep an eye on.

**Comment by birdsarah** [comment]
> How hard would it be to only output resources for models that are actually in the doc being serialized, and not simply for every Model that gets imported somewhere?

I've opened a separate issue for this already - https://github.com/bokeh/bokeh/issues/4951 - because it's pretty annoying. I think we can tackle it there.

**Comment by canavandl** [comment]
The code LGTM. Can you add a mention of resource ordering (both inter and intra-Models) to https://github.com/bokeh/bokeh/blob/master/sphinx/source/docs/user_guide/extensions.rst#supplying-external-resources?

**Comment by canavandl** [comment]
LGTM. I think the deterministic model ordering will be good, even if alphabetical isn't the

**Pull request merged**